target/linux/ath79/dts/ar7242_meraki_mr12.dts

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "ar7242.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "meraki,mr12", "qca,ar7242";
+	model = "Meraki MR12";
+
+	aliases {
+		led-boot = &led_power_orange;
+		led-failsafe = &led_power_orange;
+		led-running = &led_power_green;
+		led-upgrade = &led_power_orange;
+	};
+
+	extosc: ref {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-output-names = "ref";
+		clock-frequency = <40000000>;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		link1 {
+			label = "green:link1";
+			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
+		};
+
+		link2 {
+			label = "green:link2";
+			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
+		};
+
+		link3 {
+			label = "green:link3";
+			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+		};
+
+		link4 {
+			label = "green:link4";
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+		};
+
+		lan {
+			label = "green:lan";
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+		};
+
+		led_power_orange: power_orange {
+			label = "orange:power";
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+			panic-indicator;
+		};
+
+		led_power_green: power_green {
+			label = "green:power";
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 8 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&pcie {
+	status = "okay";
+
+	wifi@0,0,0 {
+		compatible = "pci168c,002a";
+		reg = <0x0000 0 0 0 0>;
+		qca,no-eeprom;
+		mtd-mac-address = <&config 0x66>;
+		mtd-mac-address-increment = <1>;
+	};
+};
+
+&pll {
+	clocks = <&extosc>;
+};
+
+&uart {
+	status = "okay";
+};
+
+&mdio0 {
+	status = "okay";
+
+	phy4: ethernet-phy@4 {
+		reg = <4>;
+	};
+};
+
+&eth0 {
+	status = "okay";
+	mtd-mac-address = <&config 0x66>;
+
+	pll-data = <0x02000000 0x00000101 0x00001313>;
+
+	phy-mode = "rgmii-id";
+	phy-handle = <&phy4>;
+};
+
+&mdio1 {
+	status = "okay";
+};
+
+&eth1 {
+	status = "okay";
+	mtd-mac-address = <&config 0x66>;
+};
+
+&spi {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
```

Comment:
usually 16 MB flash is capable of at least 40 MHz

Reply:
Hmm. This datasheet says this particular flash chip is capable of 50MHz: MX25L12845EMI-10G

https://html.alldatasheet.com/html-pdf/507791/MCNIX/MX25L12845EMI-10G/160/1/MX25L12845EMI-10G.html

Speaking of the specific part numbers, I need to revise the original commit to include all original parts before this can be merged.

Reply:
I have managed to boot it with a 50MHz spi-max-frequency.

I was not sure how to check directly what the SPI was running at from an initramfs, so I ran a sysupgrade against the sysupgrade squashfs I built. Over the serial interface, the standard `[e]` to `[w]` erase-write cycles seemed to move faster than the last flash. But that's not particularly scientific.

Sysupgrade worked fine. [Here](https://gist.github.com/Hurricos/d0f061ca311498552bd34343b0348078#file-mr12_second_bootlog-txt) is a gist of this boot.

Also, I checked under `/sys` to verify the `spi-max-frequency` was set:

```sh
root@OpenWrt:/sys/firmware/devicetree/base/ahb/spi@1f000000/flash@0# cat spi-max
-frequency  | hexdump
0000000 02fa f080                              
0000004
root@OpenWrt:/sys/firmware/devicetree/base/ahb/spi@1f000000/flash@0# echo $((0x0
2faf080))
50000000
```

Reply:
SPI frequency [boosted from 25MHz to 50MHz](https://github.com/openwrt/openwrt/pull/3634/commits/e783ef483365f08712c29411b2a73891e5cc2ae8#diff-ebe7e10623c0c66c5d4660984125f80f6ddb6439e5b52f53dc35dd70f48c14ceR130). Marking as resolved.